apps/designer/app/canvas/canvas.tsx

```diff
@@ -56,6 +53,101 @@ import {
   config,
 } from "~/canvas/features/wrapper-component/text-editor";
 
+// TODO: move this somewhere else
+import { useDrag } from "~/shared/design-system/components/primitives/dnd/use-drag";
+import { useDropTarget } from "~/shared/design-system/components/primitives/dnd/use-drop-target";
+import { findInstanceById, getInstancePath } from "~/shared/tree-utils";
+import { primitives } from "~/shared/canvas-components";
+const useDragAndDrop = () => {
+  const [rootInstance] = useRootInstance();
+
+  const dropTargetHandlers = useDropTarget({
+    isDropTarget(element) {
+      return (
+        rootInstance !== undefined &&
+        element.id !== "" &&
+        findInstanceById(rootInstance, element.id) !== undefined
+      );
+    },
+    onDropTargetChange(event) {
+      if (rootInstance === undefined) {
+        return;
+      }
+
+      let path = getInstancePath(rootInstance, event.target.id);
+      if (path.length === 0) {
+        path = [rootInstance];
+      }
+
+      const _targetInstance = path.reverse().find((instance) => {
+        return primitives[instance.component].canAcceptChild();
+      });
+
+      // TODO: take area into account
+      // but not sure we can use event.area, we want targetInstance's area actually
+
+      // console.log("onDropTargetChange", targetInstance);
+    },
+  });
+
+  const dragProps = useDrag({
+    onStart(_event) {
+      // const id = event.target.dataset.id;
+      // if (id == null) {
+      //   event.cancel();
+      //   return;
+      // }
+      // setDragItemId(id);
+      // autoScrollHandlers.setEnabled(true);
+    },
+    onMove: (poiterCoordinate) => {
+      dropTargetHandlers.handleMove(poiterCoordinate);
+      // autoScrollHandlers.handleMove(poiterCoordinate);
+      // placementHandlers.handleMove(poiterCoordinate);
+    },
+    onEnd() {
+      dropTargetHandlers.handleEnd();
+      // autoScrollHandlers.setEnabled(false);
+      // placementHandlers.handleEnd();
+      // setDragItemId(undefined);
+      // setPalcement(undefined);
+      // dropTargetId.current = undefined;
+    },
+  });
+
+  // We want to use <body> as a root for drag items.
+  // The DnD hooks weren't designed for that.
+  // This is a temporary solution.
+  //
+  // NOTE: maybe use root instance's element as a root?
+  useEffect(() => {
+    const handlePointerDown = (event: PointerEvent) => {
+      dragProps.onPointerDown?.(
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        event as any as React.PointerEvent<HTMLElement>
+      );
+    };
+
+    const handleScroll = () => {
+      dropTargetHandlers.handleScroll();
+    };
+
+    const rootElement = document.body;
+
+    rootElement.addEventListener("pointerdown", handlePointerDown);
```

Comment:
wasn't this supposed to be part of the primitive?

Reply:
We designed it with an assumption that the root will be available as a react element to which we can pass the props. This is not the case if we want to use `body` as root.

I'm thinking of using the root instance's node instead of body, it'll probably be more correct in any case. But even then there might not be a convenient way to add listeners to it via React. So, I'll probably change the primitive to only take refs and then add all listeners internally.